rosidl_typesupport_opensplice_c/resource/msg__type_support_c.cpp.em

```diff
@@ -51,57 +53,83 @@ extern "C"
-        header_file_name = get_header_filename_from_msg_name(field.type.type)
-        keys.add('%s/msg/%s__functions.h' % (field.type.pkg_name, header_file_name))
+        elif isinstance(type_, WString):
+            keys.add('rosidl_generator_c/u16string.h')
+            keys.add('rosidl_generator_c/u16string_functions.h')
+        elif isinstance(type_, NamespacedType):
+            if (
+                type_.name.endswith(ACTION_GOAL_SUFFIX) or
+                type_.name.endswith(ACTION_RESULT_SUFFIX) or
+                type_.name.endswith(ACTION_FEEDBACK_SUFFIX)
+            ):
+                typename = type_.name.rsplit('_', 1)[0]
+            else:
+                typename = type_.name
+            header_file_name = convert_camel_case_to_lower_case_underscore(typename)
+            keys.add('%s/%s.h' % ('/'.join(type_.namespaces), header_file_name))
     for key in keys:
         if key not in includes:
             includes[key] = set([])
-        includes[key].add(field.name)
+        includes[key].add(member.name)
 }@
 @[for key in sorted(includes.keys())]@
-#include "@(key)"  // @(', '.join(includes[key]))
+@[  if key in include_directives]@
+// already included above
+// @
+@[  else]@
+@{include_directives.add(key)}@
+@[  end if]@
+#include "@(key)"  // @(', '.join(sorted(includes[key])))
 @[end for]@
 
 // forward declare type support functions
 @{
 forward_declares = {}
-for field in spec.fields:
-    if not field.type.is_primitive_type():
-        key = (field.type.pkg_name, field.type.type)
-        if key not in includes:
+for member in message.structure.members:
+    _type = member.type
+    if isinstance(_type, NestedType):
+       _type = member.type.basetype
+
+    if isinstance(_type, NamespacedType):
+        key = (*_type.namespaces, _type.name)
+        if key not in forward_declares:
             forward_declares[key] = set([])
-        forward_declares[key].add(field.name)
+        forward_declares[key].add(member.name)
 }@
 @[for key in sorted(forward_declares.keys())]@
-@[  if key[0] != pkg]@
-ROSIDL_TYPESUPPORT_OPENSPLICE_C_IMPORT_@(pkg)
+@[  if key[0] != package_name]@
+ROSIDL_TYPESUPPORT_OPENSPLICE_C_IMPORT_@(package_name)
 @[  end if]@
 const rosidl_message_type_support_t *
-  ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(rosidl_typesupport_opensplice_c, @(key[0]), msg, @(key[1]))();
+  ROSIDL_TYPESUPPORT_INTERFACE__MESSAGE_SYMBOL_NAME(
+  rosidl_typesupport_opensplice_c, @(', '.join(key)))();
 @[end for]@
 
 @# // Make callback functions specific to this message type.
 @{
-__dds_msg_type_prefix = "{0}::{1}::dds_::{2}_".format(
-  spec.base_type.pkg_name, subfolder, spec.base_type.type)
+__dds_msg_type_prefix = '::'.join(message.structure.type.namespaces + ['dds_'] + [message.structure.type.name]) + '_'
+__ros_msg_type_prefix = '__'.join(message.structure.type.namespaces + [message.structure.type.name])
+__static_suffix = '__'.join(message.structure.type.namespaces + [message.structure.type.name])
```

Comment:
nit, could  `__static_suffix = __ros_msg_type_prefix`

Reply:
Good point, removed duplicate variable in 70f5b8c28483d6635e1d23bac087514d02d58325.